src/generators/generate-monoidal-reducer.js

```diff
@@ -0,0 +1,130 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const spec = require('shift-spec').default;
+const { sanitize, parameterize, isStatefulType } = require('../utilities.js');
+
+function reduce({ name, type }) {
+  switch (type.typeName) {
+    case 'Maybe':
+      return `${sanitize(name)} == null ? this.identity : ${sanitize(name)}`;
+    case 'List':
+      if (type.argument.typeName === 'Maybe') {
+        return `this.fold(${sanitize(name)}.filter(n => n != null))`;
+      }
+      return `this.fold(${sanitize(name)})`;
+    default:
+      return sanitize(name);
+  }
+}
+
+function reduceThunked({ name, type }) {
+  switch (type.typeName) {
+    case 'Maybe':
+      return `${sanitize(name)} == null ? () => this.identity : ${sanitize(name)}`;
+    case 'List':
+      if (type.argument.typeName === 'Maybe') {
+        return `() => this.fold(${sanitize(name)}.filter(n => n != null))`;
+      }
+      return `() => this.fold(${sanitize(name)})`;
+    default:
+      return sanitize(name);
+  }
+}
+
+function reduceForced({ name, type }) {
+  switch (type.typeName) {
+    case 'Maybe':
+      return `${sanitize(name)} == null ? this.identity : ${sanitize(name)}()`;
+    case 'List':
+      if (type.argument.typeName === 'Maybe') {
+        return `this.fold(${sanitize(name)}.filter(n => n != null))`;
+      }
+      return `this.fold(${sanitize(name)})`;
+    default:
+      return sanitize(name) + '()';
+  }
+}
+
+
+function buildContent(isThunked) {
+  let content = `/**
+ * Copyright 2016 Shape Security, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Shift from 'shift-ast';
+
+export default class ${isThunked ? 'Thunked' : ''}MonoidalReducer {
+  constructor(monoid) {
+    this.identity = monoid.empty();
+    let concat = monoid.prototype && monoid.prototype.concat || monoid.concat;
+    this.append = (a, b) => concat.call(a, b);
+  }
+
+  fold(list) {
```

Comment:
This is a breaking change since `fold` can (and should) be relied upon by consumers. I'm okay with making this change, but not until we are doing a major bump.